packages/itwin/map-layers/src/ui/FeatureInfoUiItemsProvider.tsx

```diff
@@ -32,14 +37,70 @@ export const getMapFeatureInfoToolItemDef = (): ToolItemDef =>
       await IModelApp.tools.run(MapFeatureInfoTool.toolId);
     },
     badgeType: BadgeType.TechnicalPreview,
+    isHidden: new ConditionalBooleanValue(() => {
+      // Hide the MapFeatureInfoTool if the Map Layers toggle is off or no ArcGISFeature layers are active
+      let isHidden = true;
+      if (!FeatureInfoUiItemsProvider.mapLayerProviders) {
+        FeatureInfoUiItemsProvider.loadMapLayerProviders();
+      }
+      FeatureInfoUiItemsProvider.mapLayerProviders?.forEach((provider) => {
+        if (provider?.supportsMapFeatureInfo) {
+          isHidden = false;
+          return;
+        }
+      });
+      return isHidden;
+    }, [MapLayersSyncUiEventId.MapImageryChanged, SyncUiEventId.ActiveViewportChanged, SyncUiEventId.ViewStateChanged]),
   });
 
 export class FeatureInfoUiItemsProvider implements UiItemsProvider {
   // eslint-disable-line deprecation/deprecation
   public readonly id = "FeatureInfoUiItemsProvider";
   public static readonly widgetId = "map-layers:mapFeatureInfoWidget";
+  private static _mapLayerProviders: MapLayerImageryProvider[] | undefined = undefined;
+
+  public constructor(private _featureInfoOpts: MapFeatureInfoOptions) {
+    IModelApp.viewManager.onViewOpen.addOnce((vp: ScreenViewport) => {
+      const handleMapImageryChanged = (_args: Readonly<MapImagerySettings>) => {
+        FeatureInfoUiItemsProvider.loadMapLayerProviders();
+      };
+      const handleDisplayStyleChanged = () => {
+        FeatureInfoUiItemsProvider.loadMapLayerProviders();
+      };
+      vp.displayStyle.settings.onMapImageryChanged.addListener(handleMapImageryChanged);
+      vp.onDisplayStyleChanged.addListener(handleDisplayStyleChanged);
+    });
+  }
 
-  public constructor(private _featureInfoOpts: MapFeatureInfoOptions) {}
+  /** Gets a list of currently active map layer imagery providers. */
+  public static get mapLayerProviders(): MapLayerImageryProvider[] | undefined {
+    return this._mapLayerProviders;
+  }
+
+  public static loadMapLayerProviders() {
```

Comment:
Why do we keep in memory the list of all providers being used in the view while the only information we need is a Boolean indicating if there is are least one provider supporting 'MapFeatureInfo' ?

Also we should stop iterating as soon as we encountered a first provider fulfilling our criteria.

Reply:
Resolved by refactoring code to not maintain list of providers